homeassistant/components/synology_dsm/__init__.py

```diff
@@ -144,30 +227,210 @@ async def async_setup(self):
+
+        # Reset not used API
+        if not self._with_security:
+            self.dsm.reset(self.security)
+            self.security = None
+
+        if not self._with_storage:
+            self.dsm.reset(self.storage)
+            self.storage = None
+
+        if not self._with_utilisation:
+            self.dsm.reset(self.utilisation)
+            self.utilisation = None
+
     def _fetch_device_configuration(self):
         """Fetch initial device config."""
         self.information = self.dsm.information
-        self.utilisation = self.dsm.utilisation
-        self.storage = self.dsm.storage
+
+        if self._with_security:
+            self.security = self.dsm.security
+
+        if self._with_storage:
+            self.storage = self.dsm.storage
+
+        if self._with_utilisation:
+            self.utilisation = self.dsm.utilisation
 
     async def async_unload(self):
         """Stop interacting with the NAS and prepare for removal from hass."""
         self._unsub_dispatcher()
 
-    async def update(self, now=None):
+    async def async_update(self, now=None):
         """Update function for updating API information."""
+        await self._should_fetch_api()
         await self._hass.async_add_executor_job(self.dsm.update)
         async_dispatcher_send(self._hass, self.signal_sensor_update)
+
+
+class SynologyDSMEntity(Entity):
+    """Representation of a Synology NAS entry."""
+
+    def __init__(
+        self,
+        api: SynoApi,
+        entity_type: str,
+        entity_info: Dict[str, str],
+        device_id: str = None,
+    ):
+        """Initialize the Synology DSM entity."""
+        self._api = api
+        self.entity_type = entity_type.split(":")[-1]
+        self._name = BASE_NAME
+        self._class = entity_info[ENTITY_CLASS]
+        self._enable_default = entity_info[ENTITY_ENABLE]
+        self._icon = entity_info[ENTITY_ICON]
+        self._unit = entity_info[ENTITY_UNIT]
+        self._unique_id = f"{self._api.information.serial}_{entity_type}"
+        self._device_id = device_id
+        self._device_name = None
+        self._device_manufacturer = None
+        self._device_model = None
+        self._device_firmware = None
+        self._device_type = None
+
+        if self._device_id:
+            if "volume" in entity_type:
+                volume = self._api.storage._get_volume(self._device_id)
+                # Volume does not have a name
+                self._device_name = volume["id"].replace("_", " ").capitalize()
+                self._device_manufacturer = "Synology"
+                self._device_model = self._api.information.model
+                self._device_firmware = self._api.information.version_string
+                self._device_type = (
+                    volume["device_type"]
+                    .replace("_", " ")
+                    .replace("raid", "RAID")
+                    .replace("shr", "SHR")
+                )
+            elif "disk" in entity_type:
```

Comment:
Suggest making a new class  `SynologyDSMDiskEntity`

Reply:
Did it on `SynologyDSMDeviceEntity`, if also removes double `device_info` and `available`